### PR TITLE
LPS-105705 | Create collapsible utility | v3

### PR DESCRIPTION
**Pull request by markocikos**
Hey @jbalsas ,

This is a followup on https://github.com/wincent/liferay-portal/pull/112 and https://github.com/jbalsas/liferay-portal/pull/2060

Changes:
- Collapse API is exposed in frontend-js-web
- Instead of initializing with `<aui:script require>`, we are initializing with `<liferay-frontend:component>`. It uses the newly exposed API.

Please review the code.

Thank you!

/cc @wincent

### CODE REVIEW

File: modules/apps/product-navigation/product-navigation-product-menu-web/src/main/resources/META-INF/resources/portlet/product_menu.jsp [hunk 1]
```diff
@@ -67,4 +67,16 @@
...
+
+	<%
+	Map<String, Object> context = new HashMap<>();
+
+	context.put("containerId", renderResponse.getNamespace() + "Accordion");
+	%>
+
+	<liferay-frontend:component
+		componentId='<%= renderResponse.getNamespace() + "ProductMenuCollapse" %>'
+		context="<%= context %>"
```
Comment (jbalsas) [inline modules/apps/product-navigation/product-navigation-product-menu-web/src/main/resources/META-INF/resources/portlet/product_menu.jsp]: Configuration here seems a bit of overkill, but okay, I guess... do we need `componentId`, though? are we looking to gain access to the `ProductMenuCollapse` instance?

**Comment by jbalsas** [comment]
Hey @markocikos, this looks good to me. My only concern now is... how many `*Collapse` classes and instantiations are we going to need to do? Is `ProducMenu` the sole user? How menu occurrences of `collapse` we have?

**Comment by markocikos** [comment]
> Hey @markocikos, this looks good to me. My only concern now is... how many `*Collapse` classes and instantiations are we going to need to do? Is `ProducMenu` the sole user? How menu occurrences of `collapse` we have?

In the sample case of this PR, for `product-navigation-product-menu-web`, there is only one instance of local module, as there is only one collapsable element tree.  There can be several instances in a module, for example `frontend-taglib` would have three instances.

There are total of 7 modules containing `data-toggle="collapse"`, parantheses containing usages/instances per module:
- configuration-admin-web (1)
- frontend-taglib (3)
- oauth2-provider web  (2) 
- product-navigation-control-menu-web (4)
- product-navigation-simulation-web (1)
- product-navigation-user (1)
- portal-web (2)

In addition, there are 5 usages in themes, in .ftl and .html files. I am not sure how difficult (if possible) will be to to wire those.

Sprinkled around those modules are 14 listeners on collapse events (`*.bs.collapse`)

**Comment by jbalsas** [comment]
Hey @markocikos, please see these results for `data-toggle="collapse" search:

Screen Shot 2019-12-19 at 14 46 23

I count more instances than the ones you mention... can you double check?

In any case, with this amount of occurrences and the variety of cases, I think we should attempt the auto-initialization via data-attributes to make the transition easier.

**Comment by markocikos** [comment]
> I count more instances than the ones you mention... can you double check?

I somehow missed two modules, dynamic-data-mapping-form-builder and user-associated-data-web. No idea how.

> In any case, with this amount of occurrences and the variety of cases, I think we should attempt the auto-initialization via data-attributes to make the transition easier.

So, you are ok with adding document listener to every page, like Boostrap is doing? Technically it would be every page that contains a module with dependency to frontend-js-web.

**Comment by jbalsas** [comment]
> So, you are ok with adding document listener to every page, like Boostrap is doing? Technically it would be every page that contains a module with dependency to frontend-js-web.

That's a delegate mechanism and it should be fine, yes. It should be akin to what [we did for tooltip support](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.js). We just initialize that once per page and then add the necessary delegate listeners.

Another option, if we want to keep what we have would be to just inject a small code that would do something like:

```javascript
// Pseudo code
document.querySelectorAll('[data-toggle="collapse"]').forEach(collapsible => {
    new Collapsible(collapsible);
});
```

**Comment by markocikos** [comment]
Ok, I'll rewrite.

**Comment by markocikos** [comment]
@jbalsas Should I create `frontend-js-collapse-support-web`, or place the code in `frontend-js-web`?

**Comment by jbalsas** [comment]
> @jbalsas Should I create frontend-js-collapse-support-web, or place the code in frontend-js-web?

I think I would go with that... at the risk of making the build evermore slower... `frontend-js-toggle-collapse-support-web` sounds interesting :)

**Comment by markocikos** [comment]
`frontend-js-global-web`?

**Comment by jbalsas** [comment]
> frontend-js-global-web?

No, no, the one above sounds good. Specific. We don't want to create another kitchensink for everything we don't know where else to put it :)

**Comment by markocikos** [comment]
> I think I would go with that... at the risk of making the build evermore slower... `frontend-js-toggle-collapse-support-web` sounds interesting :)

Did you mean `frontend-js-toggle-collapse-support-web` or is this a typo and you meant that `frontend-js-collapse-support-web` is ok? :)

**Comment by jbalsas** [comment]
I meant `frontend-js-toggle-collapse-support-web`, since this adds support for the `toggle="collapse"` construct. Collapse support on its own felt a bit too generic.

**Comment by jbalsas** [comment]
But I don't really care that much... it's indeed shorter without the `toggle`, so feel free to pursue one or another.

**Pull request closed without merging**